docs/internal/DistributedArchitectureGuide.md

```diff
@@ -12,6 +12,214 @@ A guide to the general Elasticsearch components can be found [here](https://gith
+ingestion, search, and everything else passes through HTTP server.
+
+Elasticsearch works over HTTP 1.1 and supports HTTP features such as: TLS,
+chunked transfer encoding, content compression, pipelining. Although we try to
+be HTTP spec compliant, Elastic is not a webserver. We support GET requests with
+payload (some old proxies might drop content), requests cannot be cached by
+middle boxes.
+
+By default HTTP server binds to the first available port in range
+9200-9300. There is no connections limit, but there is a limit on payload
+size. Default maximum payload is 100mb after compression. It's a very large
+number and almost never a good target that client should approach.
+
+Security is not enabled by default, meaning no TLS and authentication. These
+features are available in the x-pack/security module.
+
+HTTP transport provides two options for content processing: aggregate fully and
+incremental. Aggregated content is a preferable choice for a small messages that
+cannot be parsed incrementally (like JSON). But aggregation has drawbacks, it
+requires more memory, and memory is reserved until all bytes are received. If
+client sends all but the last byte, memory still in use until the client
+disconnects. If this happens concurrently then memory starts to grow
+unbounded. Large delimited content, like bulk indexing, takes advantage of
+incremental content processing, chunks of bytes. It's more complicated for
+application code, but provides better control over memory usage.
+
+Incremental bulk indexing has a back-pressure feature. When memory pressure
+grows high we stop reading bytes from TCP sockets for some connections and allow
+only few to proceed, until pressure is resolved. This feature protects from
+unbounded memory usage and OOMs.
+
+ES supports multiple Content-Types for payload: CBOR, JSON, SMILE, YAML, and
+their versioned types. Internally we call it XContentType. When a class
+implements `toXContent` that means it can be sent(serialized) over HTTP/REST.
+
+HTTP routing is based on a combination of Method and URI. For example,
+`RestCreateIndexAction` handler uses `("PUT", "/{index}")`, where curly braces
+indicate path variables. `RestBulkAction` specifies a list of routes
+
+```java
+    @Override
+    public List<Route> routes() {
+        return List.of(
+            new Route(POST, "/_bulk"),
+            new Route(PUT, "/_bulk"),
+            new Route(POST, "/{index}/_bulk"),
+            new Route(PUT, "/{index}/_bulk")
+        );
+    }
+```
+
+Every REST handler must be declared in `ActionModule` class in the
+`initRestHandlers` method. Plugins that implements `ActionPlugin` can extend
+list of handlers through `getRestHandlers` override. Every REST handler should
+extend `BaseRestHandler`.
+
+The job of the REST handler is to parse and validate HTTP request and construct
+typed version of request, often Transport request (see Transport section below).
```

Comment:
Might be worth pointing out here that (if security is enabled then) authentication happens before the REST handler but authorization happens after (when entering the transport layer).